debug: coredump: xtensa (esp32, sample_controller) support

laurenmurphyx64 opened the pull request:
Adds support for Xtensa to coredump.c and custom Python GDB server, specifically the esp32 and sample_controller SOCs.
Provides minimal set of registers shown by 'info registers'.

Try it out on qemu_xtensa:

```
west build -p -b qemu_xtensa tests/subsys/debug/coredump
west build -t run | tee coredump.log
./scripts/coredump/coredump_serial_log_parser.py coredump.log coredump.bin
./scripts/coredump/coredump_gdbserver.py build/zephyr/zephyr.elf coredump.bin
```

```
$ZEPHYR_SDK_INSTALL_DIR/xtensa/sample_controller/xtensa-zephyr-elf/bin/xtensa-zephyr-elf-gdb build/zephyr/zephyr.elf
(gdb) target remote localhost:1234
```

Try it out on esp32:

```
west build -p -b esp32 tests/subsys/debug/coredump
west flash
minicom -b 115200 -D /dev/ttyUSB1
```
Copy and paste from "***Booting..." to "E: Halting system".

Start GDB server:
```
./scripts/coredump/coredump_serial_log_parser.py coredump_esp32.log coredump_esp32.bin
./scripts/coredump/coredump_gdbserver.py build/zephyr/zephyr.elf coredump_esp32.bin 
```

Start GDB client:
```
~/.espressif/tools/zephyr/xtensa-esp32-elf/bin/xtensa-esp32-elf-gdb build/zephyr/zephyr.elf
(gdb) target remote localhost:1234
```

(Updated comment to remove issues fixed in latest force pushes or found to not be issues)

laurenmurphyx64 committed:
debug: coredump: add xtensa coredump

Adds Xtensa as supported architecture for coredump. Fixes
a few typos in documentation, Kconfig and a C file. Dumps
minimal set of registers shown by 'info registers' in GDB
for the sample_controller and ESP32 SOCs. Updates tests.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 committed:
debug: coredump: add xtensa coredump gdbserver

Adds support for Xtensa to coredump custom GDB server for
the sample_controller and ESP32 SOCs.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

nashif commented on arch/Kconfig:
```diff
@@ -104,6 +104,7 @@ config RISCV
 config XTENSA
 	bool
 	select ARCH_IS_SET
+	select ARCH_SUPPORTS_COREDUMP if SOC_XTENSA_SAMPLE_CONTROLLER || SOC_ESP32
```
we should not make this conditional on SOC, this list is going to grow and that will become unmanageable. Instead I would put this in the documentation, ie. what boards currently work.

dcpleung replied:
Or, do the `select` in the SoC layer instead.

dcpleung replied:
BTW, don't add this line in this commit until the support is actually there.

laurenmurphyx64 replied:
I combined the Kconfig changes with the PR that introduced the coredump (but not the GDB server) and added a Kconfig option `SOC_SUPPORTS_COREDUMP` selected by `SOC_ESP32` and `SOC_XTENSA_SAMPLE_CONTROLLER` and depended on by `DEBUG_COREDUMP`, let  me know what you think.

laurenmurphyx64 committed:
debug: coredump: dummy window registers, assign an to arn

GDB server dummies WINDOWBASE to 0 and WINDOWSTART to 1
for both arches and sends An as corresponding ARn for ESP32
to get around packet size limitation. Fixes backtrace and
other issues causing GDB client to crash.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 commented:
By using a suggestion from @andyross to dummy WINDOWBASE to 0 and WINDOWSTART to 1, I was able to get ESP32 SOC GDB client to calculate the An registers from corresponding ARn registers. Dummying the window registers also seemed to fix problems I was having with the sample_controller SOC. 

Although both SOCs can now backtrace and print registers, ARns are unavailable on the sample_controller. It seemed reasonable to me to not send them since I can't get the actual WINDOWBASE / WINDOWSTART to identify the actual ARs. On ESP32, the "fake" ARns (same n as An) are only set because there's no other way for me to send the Ans due to the restricted packet size. Still, if y'all think it best that I send both An and ARn on sample_controller, I can make that change.

andyross commented on arch/xtensa/core/coredump.c:
```diff
@@ -0,0 +1,159 @@
+	struct {
+		/* Minimum set shown by GDB 'info registers',
+		 * skipping user-defined register EXPSTATE
+		 *
```
For completeness, you might want to include the various MISCn (one per XCHAL_NUM_MISC_REGS) and EXCSAVEn (one for each XCHAL_NUM_INTERRUPTS).  The Zephyr kernel code uses these for various purposes, the big one being the struct cpu pointer for the current CPU.  That may take some design thought, though, because the actual register used is different for different platforms...  Future work, I guess.

andyross commented on arch/xtensa/core/coredump.c:
```diff
@@ -0,0 +1,159 @@
+#if XCHAL_HAVE_S32C1I
+	arch_blk.r.scompare1 = bsa[BSA_SCOMPARE1_OFF];
+#endif
+	arch_blk.r.sar = bsa[BSA_SAR_OFF/4];
```
In hindsight, my decision to make those constants byte offsets was probably a mistake.  Apologies.

andyross reviewed (approved):
Looks good to me, with the caveat that I'm a whole lot more comfortable with the xtensa architecture layer than I am with the gdb protocol and python scripting.

dcpleung commented on arch/xtensa/core/coredump.c:
```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
```
s/2020/2021/

dcpleung commented on soc/Kconfig:
```diff
@@ -53,3 +53,9 @@ config SOC_HAS_TIMING_FUNCTIONS
 	  Should be selected if SoC provides custom method for retrieving
 	  timestamps and cycle count.
+
+config SOC_SUPPORTS_COREDUMP
```
I don't think we need a new `SOC_SUPPORTS_*` for Xtensa. Just use the `ARCH_*` one as we know each SoC can be its own architecture for Xtensa.

laurenmurphyx64 replied:
Changes made :+1:

laurenmurphyx64 commented:
Hi, @sylvioalves, thoughts as an Espressif maintainer?

sylvioalves commented on arch/xtensa/core/coredump.c:
```diff
@@ -0,0 +1,159 @@
+#include <debug/coredump.h>
+#include <xtensa-asm2.h>
+
+#define ARCH_HDR_VER			1
```
Duplicated tab?
